felix/routetable/route_table.go

```diff
@@ -748,6 +751,78 @@ func (r *RouteTable) syncRoutesForLink(ifaceName string, fullSync bool, firstTry
 	return resyncErr
 }
 
+func (r *RouteTable) addOrOverwriteRoute(logCxt *log.Entry, nl netlinkshim.Interface, target *Target, route *netlink.Route) error {
+	err := nl.RouteAdd(route)
+	if err == nil { // Return on success
+		return nil
+	}
+	if !os.IsExist(err) || !r.shouldDeleteConflictingRoutes() {
+		return err
+	}
+
+	// There's an existing route causing a conflict.  There are a
+	// few reasons that this can happen:
+	//
+	// - We've been reconfigured and left behind some routes
+	//   that no longer appear to belong to us.  For example, the
+	//   parent device of a VXLAN tunnel changes.
+	// - Transient overlap of routes between workloads.
+	// - Overlap with an external route.
+	//
+	// Empirically, the first problem is the most common, so we try
+	// to delete the conflicting route.  The second case should
+	// "come out in the wash" when the overlap is removed.
+	// Overlap with an external route could be a problem, but it
+	// would be a misconfiguration that needs to be addressed
+	// anyway.
+	routeToDel := &netlink.Route{
+		Dst:   route.Dst,
+		Table: route.Table,
+	}
+	if conflictingRoutes, err := r.lookUpConflictingRoutes(nl, routeToDel); len(conflictingRoutes) > 0 {
+		logCxt.WithError(err).WithFields(log.Fields{
+			"calicoRoute":          target,
+			"attemptedKernelRoute": route,
+		}).Warn("Route conflicts with another, failed to look up conflicting route.")
+	} else {
+		logCxt.WithFields(log.Fields{
+			"calicoRoute":          target,
+			"attemptedKernelRoute": route,
+			"conflictingRoutes":    conflictingRoutes,
+		}).Warn("Route conflicts with another, attempting to delete conflicting route.")
+	}
+	err = nl.RouteDel(routeToDel)
+	if err != nil && !os.IsNotExist(err) {
+		logCxt.WithError(err).WithField("route", route).Warn(
+			"Failed to remove conflicting route.")
+	} else {
+		logCxt.Info("Removed conflicting route.  Retrying original add...")
+	}
+
+	return nl.RouteAdd(route)
+}
+
+func (r *RouteTable) lookUpConflictingRoutes(nl netlinkshim.Interface, route *netlink.Route) ([]netlink.Route, error) {
+	filterFlags := uint64(0)
+	if route.Table != 0 {
+		filterFlags |= netlink.RT_FILTER_TABLE
+	}
+	allRoutes, err := nl.RouteListFiltered(r.netlinkFamily, &netlink.Route{
+		Table: route.Table,
+	}, filterFlags)
+	if err != nil {
+		return nil, err
+	} else {
+		var conflictingRoutes []netlink.Route
+		for _, r := range allRoutes {
+			if r.Dst.String() == route.Dst.String() {
```

Comment:
`Dst` is of type `*net.IPNet`, so we should check for `nil` here, before calling `String()`. I recall this causing panic in some other places.